### DISCUSSION TITLE
Diagram all tables referenced in a DAX Expression

### DISCUSSION THREAD

**Discussion by edhans** [TabularEditor/TabularEditor3#1218] [Ideas]
### Is your feature request related to a problem? Please describe

I have a hornets nest of a model to go through and was trying to decipher some of the measures. I found myself using the Diagram feature to see how the data was coming together, and it is a little tedious.

### Describe the solution you'd like

Right-click on a measure, calculated column, or calculated table and have a "Diagram" feature that would throw all referenced table into the diagram. This would include other DAX expressions it was dependent on. The Dependency feature already knows all of this. Just toss that into the diagram window for me.

### Goals

1. Working through logic of a complex model
2. Debugging a particular measure or other DAX expression

**Comment by mlonsk** [comment]
Hi @edhans 

Thank you for the idea.
I am understanding you correctly that what you would like is a visual representation of the dependency tree on how different nodes of the three affect the measure?
I find it a very interesting feature, and I have a few thoughts about it that I would like your input on.

Let say we have the following measure
![image](https://github.com/TabularEditor/TabularEditor3/assets/11939114/65d05669-0f66-4546-8db9-f1d6755feba4)

How would a diagram of that make the most sense?

1. We could for example show how the tables are connected to that measures:
![image](https://github.com/TabularEditor/TabularEditor3/assets/11939114/da94f028-e266-4d5d-b179-9ad1c83a3ded)

2. We could show the dependency tree for how each table is connected to each measure:
![image](https://github.com/TabularEditor/TabularEditor3/assets/11939114/395cd99b-cde0-418b-8fac-770bc99070d7)

3. We can group the columns on a table and how they effect each measure
![image](https://github.com/TabularEditor/TabularEditor3/assets/11939114/1a04c260-db67-4593-bb56-b872e9481534)
One difficulty here is what if the tables are referenced in both measures, it might not be clear which column is used where

4.  Split out so each column is represented
![image](https://github.com/TabularEditor/TabularEditor3/assets/11939114/7471ec89-7817-47ee-b1a9-15f455dc3a15)
This might become very big and confusing if it is very complex measure

5. Columns grouped on table but still with dependency lines for each column
![image](https://github.com/TabularEditor/TabularEditor3/assets/11939114/731e5a2c-dc52-4b5e-b216-39450a530100)

These are some of the suggestions I could come up with currently, but that is also not to say that all those are technically possible, I would just like to explore what would make sense from a user perspective.

**Reply by edhans** [reply]
Wow. My initial thoughts were much more simplistic, but I like the ideas you have. The first was all I was really thinking about.

I like Number 3. Pretty simple, yet has a lot of info. The issue though becomes what if 1 table has 2 columns that affect 2 measures. If I rewrite your measure:
<img width="283" alt="image" src="https://github.com/TabularEditor/TabularEditor3/assets/45298358/c901d41b-2946-4439-88dd-d21cfd2975a8">
Then 3 becomes less clear.
<img width="657" alt="image" src="https://github.com/TabularEditor/TabularEditor3/assets/45298358/967204e6-e9af-469b-8ab4-69498ea0548f">
Where as 5, probably the most complex, shows which columns from each table affect the measure.

<img width="659" alt="image" src="https://github.com/TabularEditor/TabularEditor3/assets/45298358/41aedc18-b829-4b13-9593-cf06eaa7df8b">

I'd almost be either keep it simple with number 2, or go all out on number 5. I am not a huge fan of number 4 as it gives the impression there are three tables in play, not two.

**Reply by mlonsk** [reply]
Hi @edhans 
Thank you for the additional feedback and thoughts, much appreciated. 

I agree that there are considerations to take into account for complex measures that reference several tables, columns and measures a visual representation can quickly become quite messy. 

Even the simple second alternative becomes less clear with your measure example:
![image](https://github.com/TabularEditor/TabularEditor3/assets/11939114/d395fd15-fd46-4404-9330-bb0ce5274b15)

We will have to give this some thought internally on what makes sense both from a UX and a cognitive ease perspective. 
I imagine that the all encompassing solution is to do both option 2 and 5 and give the user the option to switch between them depending on their needs. 

In general I think the feature would be valuable, and would be really cool to also integrate into the DAX debugger to visualize those analysis even better. 

As always these things is a matter of capacity, and I can at this time not give any timeframe.

**Comment by edhans** [comment]
100% agree as long as you don't tie it _too closely_ to the debugger, as that becomes an issue until you support Calculation Groups there. 😉

**Comment by pawarbi** [comment]
I did [something similar](https://fabric.guru/measure-maze-visualizing-measure-dependencies-using-semantic-link-network-analysis) using Python, would love to see it in TE

**Comment by mlonsk** [comment]
Hi @edhans 
While this is unfortunately not something we have been able to prioritize yet, I thought I would draw your attention to Kurt's blogpost from a couple of years ago, which seems to give some of what you requested: https://data-goblins.com/power-bi/visualize-measure-dependencies

**Reply by edhans** [reply]
Thanks. I had not seen that before. Will investigate.
